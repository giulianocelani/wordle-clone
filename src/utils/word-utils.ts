import words from '../5-letters.json';

const MAX_WORD_LENGTH = 5;

function getRandomWord() {
	return words[Math.floor(Math.random() * words.length)];
}

export { getRandomWord, MAX_WORD_LENGTH };
