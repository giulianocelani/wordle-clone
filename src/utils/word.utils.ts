import words from '../5-letters.json';

const MAX_WORD_LENGTH = 5;

enum LetterState {
	MISS = 'MISS',
	MATCH = 'MATCH',
	PRESENT = 'PRESENT'
}

function getRandomWord() {
	return words[Math.floor(Math.random() * words.length)].toUpperCase();
}

function evaluateGuess(answer: string, guess: string): LetterState[] {
	if (answer.length !== guess.length && guess.length !== MAX_WORD_LENGTH) {
		return [];
	}

	const result: LetterState[] = Array(MAX_WORD_LENGTH).fill(LetterState.MISS);

	const lettersToCheck = answer.split('');
	const guessLetters = guess.split('');

	guessLetters.forEach((letter, i) => {
		if (answer[i] === letter) {
			result[i] = LetterState.MATCH;
			lettersToCheck.splice(i === lettersToCheck.length ? i - 1 : i, 1);
		}
	});

	guessLetters.forEach((letter, i) => {
		if (lettersToCheck.includes(letter) && result[i] !== LetterState.MATCH) {
			result[i] = LetterState.PRESENT;
			lettersToCheck.splice(lettersToCheck.indexOf(letter), 1);
		}
	});

	return result;
}

function isValidWord(word: string): boolean {
	return words.includes(word.toLowerCase());
}

export {
	evaluateGuess,
	getRandomWord,
	isValidWord,
	LetterState,
	MAX_WORD_LENGTH
};
