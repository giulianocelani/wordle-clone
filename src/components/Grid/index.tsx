import { getRandomWord } from '../../utils';
import Guess from '../Guess';

const MAX_GUESSES = 6;

const Grid = () => {
	const guesses = [getRandomWord()];

	return (
		<div className='flex-1 flex items-start justify-center py-10'>
			<div className='grid grid-rows-6 gap-4'>
				{[...guesses, ...Array(MAX_GUESSES - guesses.length)].map(
					(guess, index) => (
						// eslint-disable-next-line react/no-array-index-key
						<Guess key={`guess-${index}`} guess={guess} />
					)
				)}
			</div>
		</div>
	);
};

export default Grid;
