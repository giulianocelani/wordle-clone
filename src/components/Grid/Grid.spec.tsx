import { act } from 'react-dom/test-utils';

import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import useStore from '../../store';
import { LetterState } from '../../utils';

import Grid from '.';

describe('Grid', () => {
	let wrapper = mount(<Grid guess='' setGuess={jest.fn()} />);

	it('default renders correctly', () => {
		expect(toJson(wrapper, { mode: 'deep' })).toMatchSnapshot();
		expect(
			wrapper.find('div.grid-rows-6').childAt(0).props().guess.word
		).toEqual('');
	});

	it('renders only maximum 6 rows', () => {
		expect(wrapper.find('div.grid-rows-6').children().length).toBe(6);
	});

	it('renders current guess', () => {
		act(() => {
			useStore.setState({
				alreadyGuessed: [
					{
						word: 'HELLO',
						state: [
							LetterState.MISS,
							LetterState.MATCH,
							LetterState.MATCH,
							LetterState.MATCH,
							LetterState.MATCH
						]
					}
				],
				answer: 'BELLO'
			});
		});

		wrapper = mount(<Grid guess='BELL' setGuess={jest.fn()} />);
		expect(toJson(wrapper, { mode: 'deep' })).toMatchSnapshot();
		expect(
			wrapper.find('div.grid-rows-6').childAt(1).props().guess.word
		).toEqual('BELL');
	});
});
