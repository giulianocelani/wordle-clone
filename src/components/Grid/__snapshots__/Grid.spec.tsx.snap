// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid default renders correctly 1`] = `
<div
  className="flex-1 flex flex-col items-center justify-start py-5 mx-auto"
>
  <div
    className="mb-6 w-full flex space-x-2"
  >
    <input
      autoComplete="off"
      className="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5"
      disabled={false}
      maxLength={5}
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      className="text-white bg-blue-700 hover:bg-blue-800 focus:ring-4 focus:outline-none focus:ring-blue-300 font-medium rounded-lg text-sm p-2.5 text-center inline-flex items-center disabled:bg-gray-300"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <svg
        className="h-5 w-5"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clipRule="evenodd"
          d="M16.707 5.293a1 1 0 010 1.414l-8 8a1 1 0 01-1.414 0l-4-4a1 1 0 011.414-1.414L8 12.586l7.293-7.293a1 1 0 011.414 0z"
          fillRule="evenodd"
        />
      </svg>
    </button>
  </div>
  <div
    className="grid grid-rows-6 gap-4"
  >
    <div
      className="grid grid-cols-5 gap-4"
    >
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
    </div>
    <div
      className="grid grid-cols-5 gap-4"
    >
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
    </div>
    <div
      className="grid grid-cols-5 gap-4"
    >
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
    </div>
    <div
      className="grid grid-cols-5 gap-4"
    >
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
    </div>
    <div
      className="grid grid-cols-5 gap-4"
    >
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
    </div>
    <div
      className="grid grid-cols-5 gap-4"
    >
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
      <div
        className="border rounded-md flex items-center justify-center w-16 h-16 "
      >
        <h2
          className="font-semibold text-3xl "
        />
      </div>
    </div>
  </div>
</div>
`;
